server/base/src/main/java/org/apache/accumulo/server/fs/PerTableVolumeChooser.java

```
@@ -18,61 +18,146 @@
+    if (log.isTraceEnabled()) {
+      log.trace("PerTableVolumeChooser.choose");
+    }
+    VolumeChooser chooser;
+    if (!env.hasTableId() && !env.hasScope()) {
+      // Should only get here during Initialize. Configurations are not yet available.
+      return randomChooser.choose(env, options);
+    }
+
+    ServerConfigurationFactory localConf = loadConf();
+    if (env.hasScope()) {
+      // use the system configuration
+      chooser = getVolumeChooserForNonTable(env, localConf);
+    } else { // if (env.hasTableId()) {
+      // use the table configuration
+      chooser = getVolumeChooserForTable(env, localConf);
+    }
+
+    return chooser.choose(env, options);
+  }
+
+  private VolumeChooser getVolumeChooserForTable(VolumeChooserEnvironment env, ServerConfigurationFactory localConf) throws AccumuloException {
+    if (log.isTraceEnabled()) {
+      log.trace("Looking up property " + TABLE_VOLUME_CHOOSER + " for Table id: " + env.getTableId());
+    }
+    final TableConfiguration tableConf = localConf.getTableConfiguration(env.getTableId());
+    String clazz = tableConf.get(TABLE_VOLUME_CHOOSER);
+
+    return createVolumeChooser(clazz, TABLE_VOLUME_CHOOSER, env.getTableId().canonicalID(), tableSpecificChooser);
+  }
+
+  private VolumeChooser getVolumeChooserForNonTable(VolumeChooserEnvironment env, ServerConfigurationFactory localConf) throws AccumuloException {
+    String property = SCOPED_VOLUME_CHOOSER(env.getScope());
+
+    if (log.isTraceEnabled()) {
+      log.trace("Looking up property: " + property);
+    }
+
+    AccumuloConfiguration systemConfiguration = localConf.getSystemConfiguration();
+    String clazz = systemConfiguration.get(property);
+    // only if the custom property is not set to we fallback to the table volume chooser setting
+    if (null == clazz) {
+      log.debug("Property not found: " + property + " using " + DEFAULT_SCOPED_VOLUME_CHOOSER);
+      property = DEFAULT_SCOPED_VOLUME_CHOOSER;
```

I think I'm just second-guessing the basis for falling back in this way to a general-purpose "catch-all" for all scopes. It seems antithetical to the "fail hard if bad config" paradigm we discussed previously. I think there was a reason for it (backwards compatibility), but I'm not sure it's strictly necessary, since use of per-table volume choosing was always experimental.